from snakemake.utils import min_version

min_version("6.4.1")

configfile: "config/config.yaml"

# load rules
include: "rules/data.smk"
include: "rules/orthologues.smk"
include: "rules/treeinform.smk"
include: "rules/cellranger.smk"
# include: "rules/stats.smk"
# include: "rules/blastx.smk"
# include: "rules/trinotate.smk"


# target rule
rule all:
    input:
        expand("results/reference/{transcriptome}.fixed.eggnog.gtf",transcriptome=config["reference"]["filename"]),
        expand("results/cellranger/counts/{species}", species=config["species"])
        # "results/blastx-ncbi/blastx.outfmt6",
        # "results/blastx-isoseq/blastx.outfmt6"
        # expand("results/reference/{transcriptome}.transdecoder_dir/longest_orfs.pep", transcriptome=config["reference"]["filename"]),
        # expand("results/reference/{transcriptome}_longestORFperGene.pep", transcriptome=config["reference"]["filename"]),
        # expand("results/reference/{transcriptome}_longestORFperGene.fasta", transcriptome=config["reference"]["filename"]),
        # expand("results/reference/{transcriptome}_longestORFperGene.fasta.eggnog.gtf",transcriptome=config["reference"]["filename"]),
        # expand("results/cellranger-isoseq/counts/{species}", species=config["species"]),
        # "results/mitofinder",
        # "results/trinotate-isoseq/trinotate_annotation_report.xls",
        # "results/trinotate-isoseq/go_annotations.txt",
        # expand("resources/sequences/{species}.pep.fasta", species=targets.index),
        # "results/orthofinder"
