from snakemake.utils import min_version

min_version("6.4.1")

configfile: "config/config.yaml"

# load rules
# include: "rules/reference.smk"
include: "rules/cellranger.smk"
include: "rules/trinotate.smk"
include: "rules/stats.smk"
include: "rules/get_data.smk"
# include: "rules/orthologues.smk"


# target rule
rule all:
	input:
        # expand("results/reference/{transcriptome}.transdecoder_dir/longest_orfs.pep", transcriptome=config["reference"]["filename"]),
        # expand("results/reference/{transcriptome}_longestORFperGene.pep", transcriptome=config["reference"]["filename"]),
        # expand("results/reference/{transcriptome}_longestORFperGene.fasta", transcriptome=config["reference"]["filename"]),
        # expand("results/reference/{transcriptome}_longestORFperGene.fasta.eggnog.gtf",transcriptome=config["reference"]["filename"]),
        # expand("results/cellranger-isoseq/counts/{species}", species=config["species"]),
        # "results/mitofinder",
        # "results/trinotate-isoseq/trinotate_annotation_report.xls",
        # "results/trinotate-isoseq/go_annotations.txt",
		expand("resources/sequences/ensembl/{species}.pep.fasta.gz",species=ensembl_targets.loc[:,"species"]),
		expand("resources/sequences/ensemblgenomes/{species}.pep.fasta.gz",species=ensemblgenomes_targets.loc[:,"species"]),
		expand("resources/sequences/gdrive/{species}.pep.fasta",species=gdrive_targets.loc[:,"species"]),
		expand("resources/sequences/other/{species}.pep.fasta",species=other_targets.loc[:,"species"]),
		expand("resources/sequences/other_gz/{species}.pep.fasta.gz",species=other_gz_targets.loc[:,"species"])
