{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Created by Natasha Piccianis, October 12 2022\n",
    "\n",
    "\"\"\"\t\n",
    "Read a fasta file with original transcriptome,\n",
    "select the longest transcript per each gene and select corresponding lines from GTF file.\n",
    "Return one FASTA file with nucleotide sequences and one GTF file with selected transcripts.\n",
    "\"\"\"\n",
    "\n",
    "longestORFperGeneFasta = \"/Users/picciani/Dropbox/Yale/YCGA/podocoryna_analysis/reference/isoseq/Cogent_Output/Male_Med.Cogent.hq_transcripts.fasta.no5merge.collapsed.filtered.selected.rep.fa\"\n",
    "transcriptomeFile = \"/Users/picciani/Dropbox/Yale/YCGA/podocoryna_analysis/reference/isoseq/Cogent_Output/Male_Med.Cogent.hq_transcripts.fasta.no5merge.collapsed.filtered.rep.fa\"\n",
    "gff = \"/Users/picciani/Dropbox/Yale/YCGA/podocoryna_analysis/reference/isoseq/Cogent_Output/Male_Med.Cogent.hq_transcripts.fasta.no5merge.collapsed.filtered.edited.fixed.exons.corrected.gff\" # added transcript ID to a new first column; search=(.+transcript_id \"(.+)\"; gene.+)); replace=$2\\t$1\n",
    "newgtfFile = \"/Users/picciani/Dropbox/Yale/YCGA/podocoryna_analysis/reference/isoseq/Cogent_Output/Male_Med.Cogent.hq_transcripts.fasta.no5merge.collapsed.filtered.selected.edited.fixed.exons.corrected.gff\"\n",
    "identifier_type=\"type_5\"\n",
    "\n",
    "from tkinter.messagebox import YES\n",
    "import pandas as pd\n",
    "import re\n",
    "import argparse\n",
    "from pathlib import PurePosixPath\n",
    "from Bio import SeqIO\n",
    "\n",
    "def getID(header, type, whichID):\n",
    "\t\n",
    "\t\"\"\"\n",
    "\tSearch the header of a fasta file and return the gene identifier of a sequence.\"\n",
    "\n",
    "\tArguments:\n",
    "\theader -- the header string (the record.id value of a sequence when using SeqIO package).\n",
    "\ttype -- the type of gene identifier; type_1 = \"compXX_cXX_seqXX\", type_2=\"TRINITY_DNXXXXX_cX_gX_iX\",\n",
    "\t\t\ttype_3=\"SegXX.XX.XXXX\".\n",
    "\twhichID -- the kind of identifier to be retrieved; geneID = \"gene\", transcriptID = \"transcript\"\n",
    "\t\"\"\"\n",
    "\n",
    "\tif type == \"type_1\":\n",
    "\t\tsearchStr='((comp\\d+_c\\d+)_seq\\d+)'\n",
    "\tif type == \"type_2\":\n",
    "\t\tsearchStr='((TRINITY_DN\\d+_c\\d+_g\\d+)_i\\d+)'\n",
    "\tif type == \"type_3\":\n",
    "\t\tsearchStr='((Seg\\d+)\\..+)'\n",
    "\tif type == \"type_4\":\n",
    "\t\tsearchStr='(t\\d+aep.+(DN.+g.+)[_it]\\d+)'\n",
    "\tif type == \"type_5\":\n",
    "\t\tsearchStr = \"((PB.\\d+).\\d+)\"\n",
    "\tparts=header.split('\\t')\n",
    "\tp=re.match(searchStr, parts[0])\n",
    "\tgeneID=p.group(2)\n",
    "\ttranscriptID=p.group(1)\n",
    "\t\t\n",
    "\tif whichID == \"gene\":\n",
    "\t\treturn geneID\n",
    "\tif whichID == \"transcript\":\n",
    "\t\treturn transcriptID\n",
    "\n",
    "\n",
    "# # \tunique = 0 #uncomment for testing\n",
    "# # \tduplicates = 0 #uncomment for testing\n",
    "seqs={}\n",
    "transcriptNames=[]\n",
    "currentGeneID=''\n",
    "\n",
    "with open(longestORFperGeneFasta,\"w\") as outfile1:\n",
    "\tfor record in SeqIO.parse(transcriptomeFile,\"fasta\"): #find gene names\n",
    "\t\tgeneID = getID(record.id, identifier_type, whichID=\"gene\")\n",
    "\t\tif currentGeneID == '': #start the loop with an empty GeneID string\n",
    "\t\t\tcurrentGeneID = geneID\n",
    "\t\t\tseqs[record.description]=[geneID,record.seq] #add first sequence to dictionary \"seqs\", which has geneID as a list value\n",
    "\t\telse:\n",
    "\t\t\tif geneID != currentGeneID: #if the previous geneID is different from that of the current record\n",
    "# \t\t\t\t\tunique += 1 #uncomment for testing\n",
    "\t\t\t\tseqs[record.description]=[geneID,record.seq] #add unique sequences to dictionary \"seqs\"\n",
    "\t\t\t\tcurrentGeneID = geneID #re-set current geneID     \n",
    "\t\t\telse: #if the geneID is the same as that of the previous record\t\t\t\t\n",
    "# \t\t\t\t\tduplicates += 1 #uncomment for testing\n",
    "\t\t\t\tfor key, value in seqs.items(): #search for the previous record in the dictionary \"seqs\"\n",
    "\t\t\t\t\tif value[0] == currentGeneID: #and save its sequence length and name\n",
    "\t\t\t\t\t\tdupeLength=(len(value[1]))\n",
    "\t\t\t\t\t\tdupeID=key\n",
    "\t\t\t\tif dupeLength<len(record.seq): #if length of previous record is smaller that of than current record\n",
    "\t\t\t\t\tdel seqs[dupeID] #get rid of previous record\n",
    "\t\t\t\t\tseqs[record.description]=[geneID,record.seq] #add current record to dictionary\n",
    "\t\t\t\t\tcurrentGeneID = geneID\n",
    "\n",
    "\tfor name, seq in seqs.items(): #print sequences to an output file\n",
    "\t\tname = name.split(\" \")\n",
    "\t\tname = name[0]\n",
    "\t\ttranscriptNames.append(name)\n",
    "\t\tprint (\">\" + name, file=outfile1)\n",
    "\t\tprint (seq[1], file=outfile1)\n",
    "\n",
    "querylist = []\n",
    "for line in transcriptNames:\n",
    "\tline = line.strip(\"\\n\")\n",
    "\tquerylist.extend(line.split(\" \"))\n",
    "\n",
    "gtf = pd.read_table(gff, header=None)\n",
    "\n",
    "newgtf = gtf[gtf[0].isin(querylist)] \n",
    "newgtf.drop_duplicates(inplace=True)\n",
    "# #slower code to select the lines\n",
    "# for name in gtf[0]:  # for transcript name in the first column of gtf file\n",
    "# \tif (\n",
    "# \t\tname not in querylist\n",
    "# \t):  # if transcript name is not in the list of transcripts to keep, then drop it\n",
    "# \t\tindex_val = gtf[\n",
    "# \t\t\tgtf[0] == name\n",
    "# \t\t].index.values  # get the row number where it is in\n",
    "# \t\tgtf.drop(\n",
    "# \t\t\tindex=index_val, inplace=True\n",
    "# \t\t)  # drop that row and modify the actual file\n",
    "\n",
    "newgtf.to_csv(\n",
    "\tnewgtfFile, sep=\"\\t\", quoting=3, header=False, index=False\n",
    ")  # quoting=3 is the same as quoting=csv.QUOTE_NONE\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "ee0640ede16f50394a8abc0de6bd707e5c30cc01ad82abb6b38b8e894fa1504e"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 ('py39')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
